### DISCUSSION TITLE
Cost components as a non-linear function of all stops

### DISCUSSION THREAD

**Discussion by TomTruck** [google/or-tools#3307] [Routing (and legacy CP) questions]
Hi all, 

I have a quick question surrounding costs that are a non-linear function of all stops visited in a route, specifically:

**Is it possible to apply a fixed cost if _one or more_ of a certain type of stop is visited?**

For example, suppose I want to apply a fixed cost in the event that a route contains one or more stops that require a truck liftgate to perform loading/unloading.

One way to do this would be to include the vehicle in problem twice with two different cost functions, only grant access to affected stops to the vehicle that includes the special fixed cost (e.g. the liftgate cost) and finally ensure that only one is included in the solution routes.  

However, this can significantly affect run time when using cached callbacks (especially if we bifurcate each vehicle more than once).

Is there a way to do this without using the above workaround

**Comment by Mizux** [comment]
1. concerning arc cost cache, I'm not sure it will be impacted If you duplicate a vehicle, use VehicleVar to restrict the vehicles allowed to visit a node but both use the same transit callback index i.e. the cache should be share among both vehicle.

2. Another way could be to create a new dimension with a unary transit cost (like capacity dimension), each liftgate node add +1 to the CumulVar while regular node do nothing.
Vehicle with liftgate have a large capacity i.e. can visit any "liftgate" node without worry while regular vehicle have a capacity of 0 aka they can't visit these nodes...

**Comment by TomTruck** [comment]
@Mizux Thanks very much for the swift response.
